### PR TITLE
Added events

### LINKED ISSUE #5: Create app for events
### Features
* Name
* Start and End time
* Description
* Place
* Views
  * Paginated list at `/event/`
  * Single view at `/event/<id>`

Display latest 3 (should be configurable) events on the index page.

### PR DESCRIPTION
**Pull request by Prakash2403**
fixes #5

### CODE REVIEW

File: wsgi/odyssy/odyssy/urls.py [hunk 1]
```diff
@@ -18,5 +18,6 @@
 
 urlpatterns = [
     url(r'^admin/?', include(admin.site.urls)),
-    url(r'^', include('basic.urls'))
+    url(r'^events/', include('events.urls')),
```
Comment (Monal5031) [inline wsgi/odyssy/odyssy/urls.py]: I have a doubt over this regex shouldnt it be `r'^events/?'`
same with the link below

Comment (Prakash2403) [inline wsgi/odyssy/odyssy/urls.py]: Regex mentioned in the code is right. In fact, if we use $ in the end, it will cause problems. Let me explain. If this `^` is at the beginning of regex, it will match to all urls, which start with `events/.` If this `$` is at the end of regex, it will match to all urls, which end with `events/.` If combine both, i.e. `^`  at the beginning and  `$` at the end, it will match to only that urls that begin at `/events/` _and_ end at `events/`. Now, we have a requirement to give details of single event and url will be like `events/4`. If we use regex suggested by you, it will raise 404 error, because this url doesn't matches with any regex mentioned in `urls.py` file. So, we use a regex `^events/` which matches to any url which starts at `events/`.

Comment (RajuKoushik) [inline wsgi/odyssy/odyssy/urls.py]: url(r'^events/', include('events.urls')) 

This is absolutely right, no problems with this.

File: wsgi/odyssy/basic/templates/basic/index.html [hunk 2]
```diff
@@ -1,3 +1,8 @@
 <body>
     Welcome to IIIT Vadodara.
+    {% for event in events %}
+    <div class="day">
```
Comment (singhpratyush) [inline wsgi/odyssy/basic/templates/basic/index.html]: Where is this class defined?

File: wsgi/odyssy/basic/urls.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 from django.conf.urls import url
 
-import views
+from . import views
```
Comment (singhpratyush) [inline wsgi/odyssy/basic/urls.py]: Why this change?

File: wsgi/odyssy/basic/views.py [hunk 4]
```diff
@@ -1,5 +1,9 @@
+import os
+
 from django.shortcuts import render
 
+from events.models import Event
 
 def index(request):
-    return render(request, 'basic/index.html')
+    events_list = Event.objects.order_by('-event_date')[:int(os.environ.get('NUM_EVENTS_ON_INDEX_PAGE', 3 ))]
```
Comment (singhpratyush) [inline wsgi/odyssy/basic/views.py]: Do not use env var, keep it 3.

File: wsgi/odyssy/events/models.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+                    '7': 'JUL',
+                    '8': 'AUG',
+                    '9': 'SEP',
+                    '10': 'OCT',
+                    '11': 'NOV',
+                    '12': 'DEC'
+                    }
+
+    def __str__(self):
+        return self.description + " at " + self.place + " on " + \
```
Comment (singhpratyush) [inline wsgi/odyssy/events/models.py]: There are already many libraries available which can do the date to string conversion. Please use them.

File: wsgi/odyssy/events/models.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+from django.db import models
+
+
+class Event(models.Model):
+    start_time = models.DateField()
```
Comment (singhpratyush) [inline wsgi/odyssy/events/models.py]: I think this should be a `DateTimeField`. Same with the next one.

File: wsgi/odyssy/events/models.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    end_time = models.DateField()
+    event_date = models.DateField()
```
Comment (singhpratyush) [inline wsgi/odyssy/events/models.py]: This field is irrelevant.

File: wsgi/odyssy/events/models.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    description = models.CharField(max_length=50)
```
Comment (singhpratyush) [inline wsgi/odyssy/events/models.py]: Please increase it to 300.

File: wsgi/odyssy/events/templates/events/event_list.html [hunk 9]
```diff
@@ -0,0 +1,17 @@
+<body>
+    {% for event in events %}
+   <li><a href="{% url 'event-view-single' event.id %}">{{ event }}</a></li>
+    {% endfor %}
+
+    <div id="events-list-pagination">
+        <span class="step-links">
```
Comment (singhpratyush) [inline wsgi/odyssy/events/templates/events/event_list.html]: Again, where is this class defined?

File: wsgi/odyssy/events/templates/events/event_list.html [hunk 10]
```diff
@@ -0,0 +1,17 @@
+<body>
+    {% for event in events %}
+   <li><a href="{% url 'event-view-single' event.id %}">{{ event }}</a></li>
```
Comment (singhpratyush) [inline wsgi/odyssy/events/templates/events/event_list.html]: Please be consistent with link address. Use this (with `url`), or the method you used in `wsgi/odyssy/basic/templates/basic/index.html` (with `/events/...`).

File: wsgi/odyssy/events/views.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
+from django.shortcuts import get_object_or_404
+from django.core.paginator import Paginator, EmptyPage, PageNotAnInteger
+from django.shortcuts import render
+from .models import Event
+
+
+def event(request):
+    events_list = Event.objects.order_by('-event_date')
+    paginator = Paginator(events_list, 3)
```
Comment (singhpratyush) [inline wsgi/odyssy/events/views.py]: Use GET param for this number, set default to 10 instead of 3.

File: wsgi/odyssy/events/views.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    page = request.GET.get('page')
```
Comment (singhpratyush) [inline wsgi/odyssy/events/views.py]: Set default value to 1.

File: wsgi/odyssy/events/models.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+                    '8': 'AUG',
+                    '9': 'SEP',
+                    '10': 'OCT',
+                    '11': 'NOV',
+                    '12': 'DEC'
+                    }
+
+    def __str__(self):
+        return self.description + " at " + self.place + " on " + \
+               str(self.event_date.day) + " - " + self.int_to_month[str(self.event_date.month)] + \
```
Comment (arpit1997) [inline wsgi/odyssy/events/models.py]: use `datetime` library as @singhpratyush  said.

File: wsgi/odyssy/events/views.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
+from django.shortcuts import get_object_or_404
+from django.core.paginator import Paginator, EmptyPage, PageNotAnInteger
+from django.shortcuts import render
+from .models import Event
+
+
+def event(request):
```
Comment (arpit1997) [inline wsgi/odyssy/events/views.py]: use this `def event(request, page=10)`

File: wsgi/odyssy/events/views.py (hunk 12 above)
Comment (Prakash2403) [inline wsgi/odyssy/events/views.py]: Default value is set as 1. If `request.GET.get('page')` returns `None`, then it wil raise `PageNotAnInteger` Exception. This exception is catched and appropriate action is performed here.

File: wsgi/odyssy/events/models.py (hunk 7 above)
Comment (Prakash2403) [inline wsgi/odyssy/events/models.py]: I am confused with the use of `start_time` and `end_time`. As you have mentioned in a comment in #2 , I guessed that same will be applied here. This will be time period for which event description stays on-line. So, there must be some field where administrator mentions when event takes place. I am still confused, pleases suggest what to do @singhpratyush  sir ?

File: wsgi/odyssy/basic/views.py (hunk 4 above)
Comment (Prakash2403) [inline wsgi/odyssy/basic/views.py]: You have mentioned in #5  that it should be configurable, so are you sure not to use env var?

File: wsgi/odyssy/basic/urls.py (hunk 3 above)
Comment (Prakash2403) [inline wsgi/odyssy/basic/urls.py]: I don't know the reason, but it is showing some error on my machine. I tried to add current directory to PATH variable and some other methods suggested on internet, but it didn't worked. That's why I had to change. Anyone else facing same issue @Monal5031 @yashLadha ?

Comment (yashLadha) [inline wsgi/odyssy/basic/urls.py]: May be package scope is not detected, that's why you have to explicity say it to search in the current directory

File: wsgi/odyssy/events/templates/events/event_list.html (hunk 9 above)
Comment (Prakash2403) [inline wsgi/odyssy/events/templates/events/event_list.html]: Sorry sir, I was using original website to test. CSS file was included at that time. When I was creating new template, I just did copy paste. Anyways, after some time, this file is going to be change by someone, hence changing this file now is irrelevant. So, shall I change this?

File: wsgi/odyssy/basic/urls.py (hunk 3 above)
Comment (Prakash2403) [inline wsgi/odyssy/basic/urls.py]: Yes I am using pyCharm. I didn't executed it from terminal. Will check that.

Comment (singhpratyush) [inline wsgi/odyssy/basic/urls.py]: Make sure that you are using `Python2.7`.

File: wsgi/odyssy/basic/views.py (hunk 4 above)
Comment (singhpratyush) [inline wsgi/odyssy/basic/views.py]: By configurable, I meant that there should be a function that is able to call it and get results. Please create a `staticmethod` (maybe `get_latest_events`) that would return latest `n` events based on start date. This would make things easy to test and also avoid errors that would be introduced by writing it again and again.

Make such functions for everything so that we do not need to write them again while making other services.

File: wsgi/odyssy/events/models.py (hunk 7 above)
Comment (singhpratyush) [inline wsgi/odyssy/events/models.py]: The thing was there for announcements. In this case, this would imply the start and end time of the event, and not the period for which it stays on the website.

File: wsgi/odyssy/events/templates/events/event_list.html (hunk 9 above)
Comment (singhpratyush) [inline wsgi/odyssy/events/templates/events/event_list.html]: It may cause confusion in future, so I guess it would be better to remove them.

File: wsgi/odyssy/events/views.py (hunk 12 above)
Comment (singhpratyush) [inline wsgi/odyssy/events/views.py]: Though it will work, it will be more indicative to declare it here.

Comment (Prakash2403) [inline wsgi/odyssy/events/views.py]: There is no such method in `Paginator` class which accepts a default page value. I guess we will have to stick with this implementation.

File: wsgi/odyssy/events/views.py (hunk 14 above)
Comment (Prakash2403) [inline wsgi/odyssy/events/views.py]: Why to use an extra parameter here. We could obtain it from `request.GET.get()`.

**Comment by Prakash2403** [comment]
@singhpratyush @arpit1997 Please review it again

File: wsgi/odyssy/events/views.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
+from django.shortcuts import get_object_or_404
```
Comment (singhpratyush) [inline wsgi/odyssy/events/views.py]: Can you make these methods as class static methods? That way, we will just have to call `Event.get_latest_events(3)` and so on to access objects. This will ease usage. What do you think?

File: wsgi/odyssy/events/templates/events/event_list.html [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+
+    {% if events.has_next %}
+        <a href="?page={{ events.next_page_number }}&num_items={{num_items}}">Next</a>
+    {% endif %}
+
+    <a href="?num_items=4&page=1">4</a>
+    <a href="?num_items=6&page=1">6</a>
+    <a href="?num_items=8&page=1">8</a>
+    <a href="?num_items=10&page=1">10</a>
+    <a href="?num_items=20&page=1">20</a>
```
Comment (singhpratyush) [inline wsgi/odyssy/events/templates/events/event_list.html]: What do these links signify?

File: wsgi/odyssy/basic/urls.py (hunk 3 above)
Comment (singhpratyush) [inline wsgi/odyssy/basic/urls.py]: What about this one?

File: wsgi/odyssy/events/templates/events/event_list.html (hunk 16 above)
Comment (Prakash2403) [inline wsgi/odyssy/events/templates/events/event_list.html]: Number of items per page. You can select dynamically how many items you want per page

**Comment by Prakash2403** [comment]
@singhpratyush about `from . import views`. I don't think this line causes problem. Also, this is compatible with python3 and app can be launched using pyCharm. Shall I restore it to original?

**Comment by singhpratyush** [comment]
Is it compatible with `Python2`?

File: wsgi/odyssy/events/views.py (hunk 15 above)
Comment (Prakash2403) [inline wsgi/odyssy/events/views.py]: I have made such functions in views.py. But, it is better if such methods are defined in model itself. I will change it.

**Comment by Prakash2403** [comment]
Accidentally, I have uninstalled python2. Can you please check it on your machine sir?

**Comment by Prakash2403** [comment]
@singhpratyush No, it is not compatible with python2. So, which `import` style should we use in this project?

**Comment by singhpratyush** [comment]
We shall go with `python2` style as it is the version on the cloud.

File: wsgi/odyssy/basic/urls.py (hunk 3 above)
Comment (Prakash2403) [inline wsgi/odyssy/basic/urls.py]: @singhpratyush I  am facing some problems. When I am running using python 2, it is showing some error saying no module named events.apps. I am pasting whole error here.

``` (shift+enter for line break)
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/django/utils/autoreload.py", line 227, in wrapper
    fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/core/management/commands/runserver.py", line 117, in inner_run
    autoreload.raise_last_exception()
  File "/usr/local/lib/python2.7/dist-packages/django/utils/autoreload.py", line 250, in raise_last_exception
    six.reraise(*_exception)
  File "/usr/local/lib/python2.7/dist-packages/django/utils/autoreload.py", line 227, in wrapper
    fn(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/__init__.py", line 27, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python2.7/dist-packages/django/apps/registry.py", line 85, in populate
    app_config = AppConfig.create(entry)
  File "/usr/local/lib/python2.7/dist-packages/django/apps/config.py", line 120, in create
    mod = import_module(mod_path)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
ImportError: No module named events.apps

```

**Comment by Prakash2403** [comment]
@singhpratyush Please review it again.
